mixnet/src/fragment.rs

```diff
@@ -1,22 +1,249 @@
+use std::{collections::HashMap, ops::Deref};
+
+use sphinx_packet::{constants::PAYLOAD_SIZE, payload::PAYLOAD_OVERHEAD_SIZE};
+use uuid::Uuid;
+
 use crate::error::MixnetError;
 
+pub(crate) struct FragmentSet(Vec<Fragment>);
+
+const MAX_PLAIN_PAYLOAD_SIZE: usize = PAYLOAD_SIZE - PAYLOAD_OVERHEAD_SIZE;
+
+impl FragmentSet {
+    pub(crate) fn new(msg: &[u8]) -> Result<Self, MixnetError> {
+        let chunk_size = MAX_PLAIN_PAYLOAD_SIZE - FRAGMENT_HEADER_SIZE;
+
+        // For now, we don't support more than max_fragments() fragments.
+        // If needed, we can devise the FragmentSet chaining to support larger messages, like Nym.
+        let num_chunks = Self::num_chunks(msg, chunk_size);
+        if num_chunks > MAX_NUM_FRAGMENTS {
+            return Err(MixnetError::MessageTooLong(msg.len()));
+        }
+
+        let set_id = Uuid::new_v4();
+        Ok(FragmentSet(
+            msg.chunks(chunk_size)
+                .enumerate()
+                .map(|(i, chunk)| Fragment {
+                    header: FragmentHeader {
+                        set_id,
+                        total_fragments: num_chunks as FragmentId,
+                        fragment_id: i as FragmentId,
+                    },
+                    body: Vec::from(chunk),
+                })
+                .collect(),
+        ))
+    }
+
+    fn num_chunks(msg: &[u8], chunk_size: usize) -> usize {
+        msg.len() / chunk_size + (msg.len() % chunk_size > 0) as usize
+    }
+}
+
+impl Deref for FragmentSet {
+    type Target = Vec<Fragment>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
 #[derive(PartialEq, Eq, Debug, Clone)]
-pub(crate) struct Fragment {}
+pub(crate) struct Fragment {
+    header: FragmentHeader,
+    body: Vec<u8>,
+}
 
 impl Fragment {
+    pub(crate) fn bytes(&self) -> Vec<u8> {
+        let mut out = Vec::with_capacity(FRAGMENT_HEADER_SIZE + self.body.len());
+        out.extend(self.header.bytes());
+        out.extend(&self.body);
+        out
+    }
+
     pub(crate) fn from_bytes(value: &[u8]) -> Result<Self, MixnetError> {
-        todo!()
+        Ok(Self {
+            header: FragmentHeader::from_bytes(&value[0..FRAGMENT_HEADER_SIZE])?,
+            body: value[FRAGMENT_HEADER_SIZE..].to_vec(),
+        })
     }
 }
 
-pub struct MessageReconstructor {}
+type FragmentSetId = Uuid;
+type FragmentId = u8;
+const MAX_NUM_FRAGMENTS: usize = FragmentId::MAX as usize + 1;
+const FRAGMENT_HEADER_SIZE: usize = 18;
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+struct FragmentHeader {
+    set_id: FragmentSetId,
+    total_fragments: FragmentId,
+    fragment_id: FragmentId,
+}
+
+impl FragmentHeader {
+    fn bytes(&self) -> [u8; FRAGMENT_HEADER_SIZE] {
+        let mut out = [0u8; FRAGMENT_HEADER_SIZE];
+        out[0..16].copy_from_slice(self.set_id.as_bytes());
+        out[16] = self.total_fragments;
+        out[17] = self.fragment_id;
```

Comment:
This specific indexes that we use later on on different places as well may by encoded as part of the item:

```rust
impl FragmentHeader {
    const SET_ID_INDEX: usize = 16;
    ...
}
```

Then they can be reused, and if they change we can just change those instead of finding where do we need to check.